url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/engineio/client/Socket.java

```diff
@@ -0,0 +1,779 @@
+
+    /**
+     * Called after `drain` event of transport if writeBuffer is empty.
+     */
+    public static final String EVENT_DRAIN = "drain";
+
+    public static final String EVENT_HANDSHAKE = "handshake";
+    public static final String EVENT_UPGRADING = "upgrading";
+    public static final String EVENT_UPGRADE = "upgrade";
+    public static final String EVENT_PACKET = "packet";
+    public static final String EVENT_PACKET_CREATE = "packetCreate";
+    public static final String EVENT_HEARTBEAT = "heartbeat";
+    public static final String EVENT_DATA = "data";
+
+    /**
+     * Called on a new transport is created.
+     */
+    public static final String EVENT_TRANSPORT = "transport";
+
+    private static final Runnable noop = new Runnable() {
+        @Override
+        public void run() {}
+    };
+
+    /**
+     * The protocol version.
+     */
+    public static final int protocol = Parser.protocol;
+
+    public static boolean priorWebsocketSuccess = false;
+
+    private static SSLContext defaultSSLContext;
+
+    private boolean secure;
+    private boolean upgrade;
+    private boolean timestampRequests;
+    private boolean upgrading;
+    private boolean rememberUpgrade;
+    private int port;
+    private int policyPort;
+    private int prevBufferLen;
+    private long pingInterval;
+    private long pingTimeout;
+    private String id;
+    private String hostname;
+    private String path;
+    private String timestampParam;
+    private List<String> transports;
+    private List<String> upgrades;
+    private Map<String, String> query;
+    private LinkedList<Packet> writeBuffer = new LinkedList<Packet>();
+    private LinkedList<Runnable> callbackBuffer = new LinkedList<Runnable>();
+    /*package*/ Transport transport;
+    private Future pingTimeoutTimer;
+    private Future pingIntervalTimer;
+    private SSLContext sslContext;
+
+    private ReadyState readyState;
+    private ScheduledExecutorService heartbeatScheduler = Executors.newSingleThreadScheduledExecutor();
+
+    public static void setDefaultSSLContext(SSLContext sslContext) {
+        defaultSSLContext = sslContext;
+    }
+
+    public Socket() {
+        this(new Options());
+    }
+
+    /**
+     * Creates a socket.
+     *
+     * @param uri URI to connect.
+     * @throws URISyntaxException
+     */
+    public Socket(String uri) throws URISyntaxException {
+        this(uri, null);
+    }
+
+    public Socket(URI uri) {
+        this(uri, null);
+    }
+
+    /**
+     * Creates a socket with options.
+     *
+     * @param uri URI to connect.
+     * @param opts options for socket
+     * @throws URISyntaxException
+     */
+    public Socket(String uri, Options opts) throws URISyntaxException {
+        this(uri == null ? null : new URI(uri), opts);
+    }
+
+    public Socket(URI uri, Options opts) {
+        this(uri == null ? opts : Options.fromURI(uri, opts));
+    }
+
+    public Socket(Options opts) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 165-172 into a separate method.